gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,203 @@
+  }
+
+  @Override
+  public <GC_CONTENT_VALUES extends ExpiredContentValues> GCResult<GC_CONTENT_VALUES> performGC(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector) {
+    // Algorithm for performing the GC to identify the expired contents :
+
+    // Walk through each reference.
+    // For each commit in a reference, collect live table keys or dropped table keys.
+    // For each key, get the content.
+    // If the content's key is in dropped table keys,
+    //     add content to expired set (If the content is not live from previous walks of other
+    // references).
+    // Else if the commit is head commit, add contents to live set.
+    // Else check if the content is expired or not based on cutoff time of this reference.
+    //     Based on that add content to respective sets (live or expired).
+    //
+    // Note: So, If a table is dropped in a reference. Then independent of cutoff time,
+    //     all the snapshots for that table in the reference (that are not live in other reference)
+    // will be
+    //         considered for expiry.
+
+    List<Reference> references = api.getAllReferences().get().getReferences();
+    for (Reference reference : references) {
+      Instant cutOffTimestamp = this.cutOffTimestamp.apply(reference);
+      walkReference(
+          contentValuesCollector,
+          reference,
+          commitMeta ->
+              // If the commit time is older than (think: less than, before) cutoff-time, then
+              // commit is expired.
+              Objects.requireNonNull(commitMeta.getCommitTime()).compareTo(cutOffTimestamp) < 0);
+    }
+    return new GCResult<>(contentValuesCollector.contentValues);
+  }
+
+  private <GC_CONTENT_VALUES extends ExpiredContentValues> void walkReference(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      Predicate<CommitMeta> expiredCommitPredicate) {
+    try (Stream<LogResponse.LogEntry> commits =
+        StreamingUtil.getCommitLogStream(
+            api, reference.getName(), null, null, null, OptionalInt.empty(), true)) {
+
+      Set<ContentKey> deletedKeys = new HashSet<>();
+      AtomicBoolean commitHead = new AtomicBoolean(true);
+      commits.forEach(
+          (logEntry) -> {
+            boolean isExpired;
+            if (commitHead.get()) {
+              // If the commit is head commit, then it should be kept live if the table is reachable
+              isExpired = false;
+              commitHead.set(false);
+            } else {
+              // check expiry based on cutoff time
+              isExpired = expiredCommitPredicate.test(logEntry.getCommitMeta());
+            }
+
+            if (logEntry.getOperations() != null) {
+              logEntry
+                  .getOperations()
+                  .forEach(
+                      operation -> {
+                        if (operation instanceof ImmutableDelete) {
```

Comment:
I am not sure why we need to list the operations? additionally, what if we add another type of te `Delete` interface?

Reply:
this is to collect the drop table's keys

Reply:
changed to `Operation.Delete` instead of `ImmutableDelete`